core/src/main/java/org/infinispan/encoding/ProtostreamTranscoder.java

```diff
@@ -132,27 +124,92 @@ private boolean isWrapped(MediaType mediaType) {
    }
 
    private byte[] marshall(Object decoded, MediaType destinationType) throws IOException {
-      try {
-         if (isWrapped(destinationType)) return ProtobufUtil.toWrappedByteArray(ctx(), decoded);
-         return ProtobufUtil.toByteArray(ctx(), decoded);
-      } catch (IllegalArgumentException iae) {
-         throw new MarshallingException(iae.getMessage());
+      ImmutableSerializationContext ctx = getCtxForMarshalling(decoded);
+      if (isWrapped(destinationType)) {
+         return ProtobufUtil.toWrappedByteArray(ctx, decoded);
       }
+      return ProtobufUtil.toByteArray(ctx, decoded);
    }
 
    private Object unmarshall(byte[] bytes, MediaType contentType, MediaType destinationType) throws IOException {
+      if (isWrapped(contentType))
+         return unmarshallCascading(bytes);
+
+      String type = destinationType.getClassType();
+      if (type == null) throw logger.missingTypeForUnwrappedPayload();
+      Class<?> destination = Util.loadClass(type, classLoader);
+      ImmutableSerializationContext ctx = getCtxForMarshalling(destination);
+      return ProtobufUtil.fromByteArray(ctx, bytes, destination);
+   }
+
+   private Object unmarshallCascading(byte[] bytes) throws IOException {
+      // First try to unmarshalling with the user context
```

Comment:
Does it have a JIRA to have this workaround fixed in the protostream side?

Reply:
Yes [IPROTO-139](https://issues.redhat.com/browse/IPROTO-139). I'll add comments to the workaround methods.